src/Microsoft.ML.Core/Data/SchemaBuilder.cs

```diff
@@ -0,0 +1,82 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using Microsoft.ML.Runtime.Data;
+using System;
+using System.Collections.Generic;
+using System.Text;
+
+namespace Microsoft.ML.Data
+{
+    /// <summary>
+    /// A builder for <see cref="Schema"/>.
+    /// </summary>
+    public sealed class SchemaBuilder
+    {
+        private readonly List<(string Name, ColumnType Type, Schema.Metadata Metadata)> _items;
+
+        /// <summary>
+        /// Create a new instance of <see cref="SchemaBuilder"/>.
+        /// </summary>
+        public SchemaBuilder()
+        {
+            _items = new List<(string Name, ColumnType Type, Schema.Metadata Metadata)>();
+        }
+
+        /// <summary>
+        /// Add one column to the schema being built.
+        /// </summary>
+        /// <param name="name">The column name.</param>
+        /// <param name="type">The column type.</param>
+        /// <param name="metadata">The column metadata.</param>
+        public void AddColumn(string name, ColumnType type, Schema.Metadata metadata)
```

Comment:
>Schema.Metadata metadata [](start = 60, length = 24)

Do you wanna have override method which accept metadata builder instead? 
Can we even have wild metadata by itself running on lush fields of schema and not be tamed by MetadataBuilder?

Reply:
I don't like making one builder take another builder as a parameter, it introduces ambiguity:
For instance, what if I do
```csharp
metadataBuilder.AddSlotNames(getter);
schemaBuilder.AddColumn("Foo", columnType, metadataBuilder);
metadataBuilder.AddKeyValues(anotherGetter);
var result = schemaBuilder.GetSchema();
```
will the result have key velues metadata or not?

I'd rather not even allow for this ambiguity.

---
In reply to: [235486545](https://github.com/dotnet/machinelearning/pull/1680#discussion_r235486545) [](ancestors = 235486545)